walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,68 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Set;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // if there's no requested attendee, whole day is available
+    if(request.getAttendees().isEmpty()){
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // nobody has time to meet for more than a day
+    if(request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return Arrays.asList();
+    }
+    // if there're no already existing events, whole day is available, 
+    if(events.isEmpty()){
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<String> listOfAttendees = request.getAttendees();
+    List<TimeRange> unavailableTimes = getUnavailableTimesForAttendees(events, listOfAttendees);
+
+    return getAvailableTimesForAttendees(unavailableTimes);
+  }
+  // returns a list of already existing commitments for all required attendees
+  public List<TimeRange> getUnavailableTimesForAttendees(Collection<Event> events, Collection<String> listOfAttendees){
+    List<TimeRange> unavailableTimes = new ArrayList<>();
+    for(String attendee : listOfAttendees) {
+      for(Event event : events) {
+        Set<String> eventAttendees = event.getAttendees();
+        if(eventAttendees.contains(attendee)) {
+          unavailableTimes.add(event.getWhen());
+        }
+      }
+    }
+    return unavailableTimes;
+  }
+
+  // returns all available times given already existing commitments
+  public Collection<TimeRange> getAvailableTimesForAttendees(List<TimeRange> unavailableTimes){
+    int startTime = TimeRange.START_OF_DAY;
+    Collections.sort(unavailableTimes, TimeRange.ORDER_BY_START);
+    Collection<TimeRange> availableTimes = new ArrayList<>();
+    
+    for(TimeRange unavailableTime : unavailableTimes) {
+      if(startTime < unavailableTime.start()) {
```

Comment:
Sorry if I'm misunderstanding something, but won't this add times even if they're too short? Like in this case:

Meeting request duration: 1 hour
First unavailable time: 00:30 - 01:00

Won't this loop add 00:00 - 00:30 as the first available time, even though it's less than the requested duration?

Reply:
You're right. I'll update for that.

Reply:
Nice, thanks for looking into this!